cpp/src/arrow/util/decimal.h

```diff
@@ -61,6 +63,23 @@ class ARROW_EXPORT Decimal128 : public BasicDecimal128 {
   // This is required on some older compilers.
   constexpr Decimal128() noexcept : BasicDecimal128() {}
 
+  /// Divide this number by right and return the result.
+  ///
+  /// This operation is not destructive.
+  /// The answer rounds to zero. Signs work like:
+  ///   21 /  5 ->  4,  1
+  ///  -21 /  5 -> -4, -1
+  ///   21 / -5 -> -4,  1
+  ///  -21 / -5 ->  4, -1
+  /// \param[in] divisor the number to divide by
+  /// \return the pair of the quotient and the remainder
+  Result<std::pair<Decimal128, Decimal128>> Divide(const Decimal128& divisor) const {
+    std::pair<Decimal128, Decimal128> result;
+    auto dstatus = BasicDecimal128::Divide(divisor, &result.first, &result.second);
```

Comment:
Why not also migrate BasicDecimal128?

Reply:
To migrate `BasicDecimal128` APIs, we need to convert the value of `DecimalStatus` to `arrow::Status`.  But this conversion makes a loss of information because the values except `kSuccess` are converted to `Status::Invalid`.

DecimalStatus is very clear to represent the result status of decimal operations, so I think we should keep `BasicDecimal128` APIs as it is.

Reply:
I see, thank you.